fastjet/internal/delaunay/voronoi.go

```diff
@@ -0,0 +1,343 @@
+// Copyright 2017 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package delaunay
+
+import (
+	"fmt"
+	"math"
+)
+
+// Voronoi holds the border information
+type Voronoi struct {
+	maxX float64
+	minX float64
+	maxY float64
+	minY float64
+}
+
+// FIXME can't do any delaunay operation after calling NewVoronoi
+func NewVoronoi(d *Delaunay) *Voronoi {
+	border := make([]*Triangle, 0, len(d.root.A.adjacentTriangles)+len(d.root.B.adjacentTriangles)+len(d.root.C.adjacentTriangles))
+	border = append(border, d.root.A.adjacentTriangles[:]...)
+	border = append(border, d.root.B.adjacentTriangles[:]...)
+	border = append(border, d.root.C.adjacentTriangles[:]...)
+	for i, t := range d.triangles {
+		if len(d.triangles[i].children) == 0 {
+			for j, tri := range border {
+				if tri.Equals(t) {
+					border = append(border[:j], border[j+1:]...)
+					t.A.isOutside = true
+					if !t.A.Equals(d.root.A) && !t.A.Equals(d.root.B) && !t.A.Equals(d.root.C) {
+						t.A.adjacentTriangles = remove(t.A.adjacentTriangles, t)
+					}
+					t.B.isOutside = true
+					if !t.B.Equals(d.root.A) && !t.B.Equals(d.root.B) && !t.B.Equals(d.root.C) {
+						t.B.adjacentTriangles = remove(t.B.adjacentTriangles, t)
+					}
+					t.C.isOutside = true
+					if !t.C.Equals(d.root.A) && !t.C.Equals(d.root.B) && !t.C.Equals(d.root.C) {
+						t.C.adjacentTriangles = remove(t.C.adjacentTriangles, t)
+					}
+					break
+				}
+			}
+		}
+	}
+	v := &Voronoi{maxX: d.maxX, minX: d.minX, maxY: d.maxY, minY: d.minY}
+	return v
+}
+
+// finds the voronoi area of a point
+// returns the area and the points associated with it
+func (v *Voronoi) VoronoiCell(p *Point) (float64, []*Point) {
+	// find all points that form the voronoi cell for that point
+	// in clockwise order
+	centers := make([]*Point, len(p.adjacentTriangles), len(p.adjacentTriangles)+3)
+	var t, first *Triangle
+	if p.isOutside {
+		// find first triangle on the outside
+		for _, t1 := range p.adjacentTriangles {
+			var pt *Point
+			if t1.A.Equals(p) {
+				pt = t1.C
+			} else if t1.B.Equals(p) {
+				pt = t1.A
+			} else if t1.C.Equals(p) {
+				pt = t1.B
+			} else {
+				// should never happen
+				panic(fmt.Errorf("voronoi: internal error with adjacent triangles for point P%s", p))
+			}
+			border := true
+			for _, t2 := range p.adjacentTriangles {
+				if t1.Equals(t2) {
+					continue
+				}
+				if pt.Equals(t2.A) || pt.Equals(t2.B) || pt.Equals(t2.C) {
+					border = false
+				}
+			}
+			if border {
+				t = t1
+				break
+			}
+		}
+		first = t
+		if t == nil {
+			// should only happen when user makes a mistake
```

Comment:
What kind of mistake? The panic message should be more informative, so they can understand what they did. Then remove the comment, since the code will explain itself.